File: ina_1sem/labor/5_lab/5_task/findzero.py
from math import cos

def find_zero(f, a, b, eps):
    if f(a) * f(b) >= 0:
        print("Error: the function must have different signs at the ends of the interval.")
        return None

    c, prev_c = a, a
    while (b - a) / 2.0 > eps:
        c = (a + b) / 2.0
        if f(c) == 0.0:
            return c
        elif f(a) * f(c) < 0:
            b = c
        else:
            a = c

        if abs(c - prev_c) < eps:
            break
        prev_c = c

    return c

def test_function(x):
    return cos(x / 2)

def main():
    a, b = 2.0, 4.0
    for i in range(1, 9):
        eps = 10 ** -i
        zero = find_zero(test_function, a, b, eps)
        if zero is not None:
            format_string = f"Zero point: {{:.{i}f}} (found with precision ε = {{:.{i}f}})"
            print(format_string.format(zero, eps))

if __name__ == "__main__":
    main()

File: ina_1sem/labor/5_lab/3_task/findzero.c
#include <stdio.h>
#include <math.h>

typedef double (*functype)(double);

double findzero(functype f, double a, double b, double eps) {
    if (f(a) * f(b) >= 0) {
        printf("Error: the function must have different signs at the ends of the interval.\n");
        return NAN;
    }

    double c, prev_c = a;
    do {
        c = (a + b) / 2.0;
        if (f(c) == 0.0) {
            return c;
        } else if (f(a) * f(c) < 0) {
            b = c;
        } else {
            a = c;
        }

        if (fabs(c - prev_c) < eps) {
            break;
        }
        prev_c = c;
    } while ((b - a) / 2.0 > eps);

    return c;
}

double test_function(double x) {
    return cos(x / 2);
}

int main() {
    double a = 2.0, b = 4.0;
    double eps;
    double zero;

    for (int i = 1; i <= 8; i++) {
        eps = pow(0.1, (double)i);
        zero = findzero(test_function, a, b, eps);

        if (!isnan(zero)) {
            char format[100];
            sprintf(format, "Zero point: %%.%df (found with precision ε = %%.%df)\n", i, i);
            printf(format, zero, eps);
        }
    }

    return 0;
}
